### PR TITLE
Added 'Show original page' link into distilled site.

### PR DESCRIPTION
**Pull request by boocmp**
Resolves https://github.com/brave/brave-browser/issues/23445

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/speedreader/renderer/speedreader_render_frame_observer.cc
Comment (iefremov) [inline components/speedreader/renderer/speedreader_render_frame_observer.cc:10]: pls check for unused includes everywhere

File: components/speedreader/renderer/speedreader_js_handler.h
Comment (iefremov) [inline components/speedreader/renderer/speedreader_js_handler.h:16]: not used

File: components/speedreader/renderer/DEPS
Comment (iefremov) [inline components/speedreader/renderer/DEPS:6]: ui ?..

File: components/speedreader/common/speedreader.mojom
Comment (iefremov) [inline components/speedreader/common/speedreader.mojom:9]: comment

File: browser/speedreader/speedreader_tab_helper.h
Comment (iefremov) [inline browser/speedreader/speedreader_tab_helper.h:142]: comments

File: browser/speedreader/speedreader_tab_helper.h
Comment (iefremov) [inline browser/speedreader/speedreader_tab_helper.h:42]: comments on why would we want to do that

**Comment by iefremov** [comment]
@diracdeltas can you please take a quick look - do we need an official security review for this? We are injecting a JS api and a link into speedreader-distilled pages to reload the original page.

**Comment by iefremov** [comment]
@boocmp we can consider gin templates to reduce some boilerplate https://github.com/brave/brave-core/pull/13645/files#diff-6b746d64042e5ea6c6673b1bba6e4f9a99035c919171ff0d9d63e65906c60869R312

File: browser/speedreader/speedreader_tab_helper.cc
Comment (diracdeltas) [inline browser/speedreader/speedreader_tab_helper.cc:373]: just wondering does this page have a content security policy that would prevent inline JS execution? if so i assume this kind of injected JS is exempt from the policy?

File: browser/speedreader/speedreader_tab_helper.cc
Comment (diracdeltas) [inline browser/speedreader/speedreader_tab_helper.cc:360]: there should be some sanitization here; otherwise if a translator could do a JS injection by setting `$1` to be `'; someJavaScript(); '`. i believe there are some scripts which check that our translations don't contain unsafe HTML but it wouldn't catch javascript injection.

File: components/speedreader/renderer/speedreader_js_handler.cc
Comment (diracdeltas) [inline components/speedreader/renderer/speedreader_js_handler.cc:66]: you probably want to set all these objects to be non-configurable and non-writable, like we do for `window.ethereum`

File: components/speedreader/renderer/speedreader_render_frame_observer.cc
Comment (diracdeltas) [inline components/speedreader/renderer/speedreader_render_frame_observer.cc:26]: should maybe check here if the URL scheme is in a whitelisted set, ex: `http, https`

File: browser/speedreader/speedreader_tab_helper.cc
Comment (diracdeltas) [inline browser/speedreader/speedreader_tab_helper.cc:352]: just making sure i understand: this will run in the same isolated world as Translate? https://source.chromium.org/chromium/chromium/src/+/main:chrome/common/chrome_isolated_world_ids.h;l=13?q=ISOLATED_WORLD_ID_CONTENT_END&ss=chromium

File: browser/speedreader/speedreader_tab_helper.cc
Comment (diracdeltas) [reply browser/speedreader/speedreader_tab_helper.cc:373]: you can make a test for this by adding an html page that contains `<head><meta http-equiv="Content-Security-Policy" content="default-src 'self'"></head>` and seeing if the link gets injected

File: browser/speedreader/speedreader_tab_helper.cc
Comment (diracdeltas) [inline browser/speedreader/speedreader_tab_helper.cc:362]: i assume js on the page isn't able to call `window.speedreader` since it's in an isolated world, but could page js call `document.getElementById('show-original-page').click()` ? just wondering if a page can abuse this feature to force the user to always see the original page.

File: components/speedreader/rust/lib/src/readability/src/extractor.rs
Comment (diracdeltas) [inline components/speedreader/rust/lib/src/readability/src/extractor.rs:334]: consider calling the ID something more unique in case the page already has an element with ID `show-original-page`

**Comment by diracdeltas** [comment]
i wonder if it'd be better to put the "show original page" link in the browser UI somewhere instead of on the page. can the page hide this element with a CSS rule for instance?

File: browser/speedreader/speedreader_tab_helper.cc
Comment (diracdeltas) [reply browser/speedreader/speedreader_tab_helper.cc:362]: NVM i am told that pages in speedreader mode can't run JS

File: browser/speedreader/speedreader_tab_helper.cc
Comment (boocmp) [reply browser/speedreader/speedreader_tab_helper.cc:373]: we have this CSP in speedreader: https://github.com/brave/brave-core/blob/05f2a4191f0dd1a50d62e7a40c9fb9506fe53f57/components/speedreader/speedreader_rewriter_service.cc#L36

File: browser/speedreader/speedreader_tab_helper.cc
Comment (boocmp) [reply browser/speedreader/speedreader_tab_helper.cc:352]: Yep, but it i can change it if needed.

**Review by simonhong — DISMISSED** [review]
++ for `browser/themes` trivial changes.

File: browser/speedreader/speedreader_tab_helper.cc
Comment (diracdeltas) [reply browser/speedreader/speedreader_tab_helper.cc:352]: probably best to use a new world

**Comment by iefremov** [comment]
@simonhong sorry it needs your approval again

File: browser/speedreader/speedreader_tab_helper.cc
Comment (diracdeltas) [inline browser/speedreader/speedreader_tab_helper.cc:371]: maybe `\u0027` could still be used to inject `'`

could just add another CHECK that it doesn't contain `\`

Comment (boocmp) [reply browser/speedreader/speedreader_tab_helper.cc:371]: Done

**Review by diracdeltas — APPROVED** [review]

**Review by simonhong — APPROVED** [review]

**Pull request merged**
